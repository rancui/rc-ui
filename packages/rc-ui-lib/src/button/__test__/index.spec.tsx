import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import Icon from '../../icon';
import Button from '..';

describe('Button', () => {
  it('renders correctly', () => {
    expect(<Button />).toMatchRenderedSnapshot();
  });

  it('mount correctly', () => {
    expect(() => render(<Button />)).not.toThrow();
  });

  it('should render empty button without errors', () => {
    const { container } = render(
      <Button>
        {null}
        {undefined}
      </Button>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should simulate click event', async () => {
    const onClick = jest.fn();

    const { container } = render(<Button onClick={onClick} plain />);

    await fireEvent.click(container.querySelector('.rc-button'));

    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it('should not emit click event when disabled', async () => {
    const onClick = jest.fn();
    const { container } = render(<Button disabled onClick={onClick} />);
    await fireEvent.click(container.querySelector('.rc-button'));
    expect(onClick).not.toHaveBeenCalled();
  });

  it('should render correctly when loading prop is set', () => {
    const { container, rerender } = render(<Button loading />);
    expect(container.querySelector('.rc-loading__spinner')).toBeTruthy();

    rerender(<Button type="primary" loading />);
    expect(
      container.querySelector('.rc-button').classList.contains('rc-button--primary'),
    ).toBeTruthy();
  });

  it('should not emit click event when loading prop is set', async () => {
    const onClick = jest.fn();
    const { container } = render(<Button loading onClick={onClick} />);
    await fireEvent.click(container.querySelector('.rc-button'));
    expect(onClick).not.toHaveBeenCalled();
  });

  it('should hide border when color is gradient', () => {
    const { container, rerender } = render(<Button color="linear-gradient(#000, #fff)" />);
    const style = getComputedStyle(container.querySelector('.rc-button'));
    expect(style.border).toEqual('0px');

    rerender(<Button color="linear-gradient(#000, #fff)" plain />);
    const style2 = getComputedStyle(container.querySelector('.rc-button'));
    expect(style2.color).toEqual('rgb(255, 255, 255)');
    expect(
      container.querySelector('.rc-button').classList.contains('rc-button--plain'),
    ).toBeTruthy();
  });

  it('should hide border when color is not gradient', () => {
    const { container } = render(<Button type="primary" color="red" />);
    const style = getComputedStyle(container.querySelector('.rc-button'));
    expect(style.borderColor).toEqual('red');
  });

  it('should change icon class prefix when using icon-prefix prop', () => {
    const { container } = render(<Button icon="success" iconPrefix="my-icon" />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly when icon is JSX element', () => {
    const { container } = render(<Button icon={<Icon name="shop-o" />} />);
    expect(container).toMatchSnapshot();
  });

  it('should render iconPosition correctly', () => {
    const { container } = render(<Button type="primary" iconPosition="right" />);
    expect(container.querySelector('.rc-loading__spinner')).toBeFalsy();
  });

  it('should render buttton group correctly', () => {
    const { container } = render(
      <Button.Group>
        <Button type="primary">next step</Button>
      </Button.Group>,
    );
    expect(container.querySelector('.rc-button-group')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
});
