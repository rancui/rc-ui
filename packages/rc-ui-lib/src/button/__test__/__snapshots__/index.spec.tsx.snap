// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<div>
  <button
    class="rc-button rc-button--default rc-button--normal"
  >
    <div
      class="rc-button__content"
    />
  </button>
</div>
`;

exports[`Button should change icon class prefix when using icon-prefix prop 1`] = `
<div>
  <button
    class="rc-button rc-button--default rc-button--normal"
  >
    <div
      class="rc-button__content"
    >
      <i
        class="rc-badge__wrapper rc-button__icon my-icon my-icon-success"
      />
    </div>
  </button>
</div>
`;

exports[`Button should render buttton group correctly 1`] = `
<div>
  <div
    class="rc-button-group"
  >
    <button
      class="rc-button rc-button--primary rc-button--normal"
    >
      <div
        class="rc-button__content"
      >
        <span
          class="rc-button__text"
        >
          next step
        </span>
      </div>
    </button>
  </div>
</div>
`;

exports[`Button should render correctly when icon is JSX element 1`] = `
<div>
  <button
    class="rc-button rc-button--default rc-button--normal"
  >
    <div
      class="rc-button__content"
    >
      <i
        class="rc-badge__wrapper van-icon van-icon-shop-o"
      />
    </div>
  </button>
</div>
`;

exports[`Button should render empty button without errors 1`] = `
<div>
  <button
    class="rc-button rc-button--default rc-button--normal"
  >
    <div
      class="rc-button__content"
    >
      <span
        class="rc-button__text"
      />
    </div>
  </button>
</div>
`;
